WordPress/Classes/Services/MediaImportService.swift

```diff
@@ -20,55 +20,254 @@ open class MediaImportService: LocalCoreDataService {
+                    // Pre-generate a thumbnail image, see the method notes.
+                    self.exportPlaceholderThumbnail(for: media) { url in
+                        assert(Thread.isMainThread)
+                        guard let url, let media = try? self.coreDataStack.mainContext.existingObject(with: media.objectID) as? Media else {
+                            return
+                        }
+                        thumbnailCallback?(media, url)
+                    }
+                }
+
+                completion(transformed)
+            }, on: .main)
+        }
+
+        return self.import(exportable, to: media, allowableFileExtensions: allowedFileTypes, completion: completion)
+    }
+
     /// Imports media from a PHAsset to the Media object, asynchronously.
     ///
     /// - Parameters:
     ///     - exportable: the exportable resource where data will be read from.
     ///     - media: the media object to where media will be imported to.
-    ///     - onCompletion: Called if the Media was successfully created and the asset's data imported to the absoluteLocalURL.
-    ///     - onError: Called if an error was encountered during creation, error convertible to NSError with a localized description.
+    ///     - onCompletion: Called if the Media was successfully created and the asset's data imported to the
+    ///         absoluteLocalURL. This closure is called on the main thread. The closure's `media` argument is also
+    ///         bound to the main context (`CoreDataStack.mainContext`).
+    ///     - onError: Called if an error was encountered during creation, error convertible to NSError with a
+    ///         localized description. This closure is called on the main thread.
     ///
     /// - Returns: a progress object that report the current state of the import process.
     ///
-    @objc(importResource:toMedia:onCompletion:onError:)
-    func `import`(_ exportable: ExportableAsset, to media: Media, onCompletion: @escaping MediaCompletion, onError: @escaping OnError) -> Progress? {
+    private func `import`(_ exportable: ExportableAsset, to media: Media, allowableFileExtensions: Set<String>, completion: @escaping (Result<Media, Error>) -> Void) -> Progress {
         let progress: Progress = Progress.discreteProgress(totalUnitCount: 1)
         importQueue.async {
-            guard let exporter = self.makeExporter(for: exportable) else {
+            guard let exporter = self.makeExporter(for: exportable, allowableFileExtensions: allowableFileExtensions) else {
                 preconditionFailure("An exporter needs to be availale")
             }
-            let exportProgress = exporter.export(onCompletion: { export in
-                self.managedObjectContext.perform {
-                    self.configureMedia(media, withExport: export)
-                    ContextManager.sharedInstance().save(self.managedObjectContext, completion: {
-                        onCompletion(media)
```

Comment:
⚠️ Potential issue: `media` is changed in the `managedObjectContext` queue, but there is no guarantee it's bound to the context object.

Reply:
⚠️ Potential issue: `media` is send to the `onCompletion` closure on the main thread, but we don't know if `media` is bound to the main context.